jupyterhub/apihandlers/base.py

```diff
@@ -257,34 +286,36 @@ def user_model(self, user, include_servers=False, include_state=False):
         if model:
             if '' in user.spawners and 'pending' in allowed_keys:
                 model['pending'] = user.spawners[''].pending
-            if include_servers and 'servers' in allowed_keys:
-                # Todo: Replace include_state with scope (read|admin):users:auth_state
+            if 'servers' in allowed_keys:
                 servers = model['servers'] = {}
                 for name, spawner in user.spawners.items():
                     # include 'active' servers, not just ready
                     # (this includes pending events)
                     if spawner.active:
-                        servers[name] = self.server_model(
-                            spawner, include_state=include_state
-                        )
+                        servers[name] = self.server_model(spawner)
         return model
 
-    def group_model(self, group):  # Todo: make consistent to do scope checking here
+    def group_model(self, group):
         """Get the JSON model for a Group object"""
-        return {
-            'kind': 'group',
-            'name': group.name,
-            'users': [u.name for u in group.users],
-        }
+        model = {'kind': 'group', 'name': group.name}
+        req_scope = 'read:groups'
+        if req_scope in self.parsed_scopes:
+            scope_filter = self.get_scope_filter(req_scope)
+            if scope_filter(group, kind='group'):
+                model['users'] = [u.name for u in group.users]
+        return model
 
-    def service_model(self, service):  # Todo: make consistent to do scope checking here
+    def service_model(self, service):
         """Get the JSON model for a Service object"""
-        return {
-            'kind': 'service',
-            'name': service.name,
-            'admin': service.admin,
-            'roles': [r.name for r in service.roles],
-        }
+        model = {'kind': 'service', 'name': service.name}
+        req_scope = 'read:services'
+        if req_scope in self.parsed_scopes:
+            scope_filter = self.get_scope_filter(req_scope)
+            if scope_filter(service, kind='service'):
```

Comment:
If this is False, the kind and name are both still returned. Should the service model be entirely empty if there is no read access to this service?